### PR TITLE
Implements #5939

### PR DESCRIPTION
**Pull request by hmic**
Implements https://github.com/cakephp/cakephp/issues/5939

### CODE REVIEW

File: src/I18n/Date.php [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+     * @see \Cake\I18n\Time::nice()
+     */
+    public static $niceFormat = [IntlDateFormatter::MEDIUM, -1];
+
+    /**
+     * The default locale to be used for displaying formatted date strings.
+     *
+     * @var string
+     */
+    public static $defaultLocale;
```
Comment (ADmad) [inline src/I18n/Date.php]: Nothing overridden here so doesn't seem necessary.

Comment (hmic) [inline src/I18n/Date.php]: True, so true...

It's late here already.

**Comment by hmic** [comment]
Why is travis failing in random and unrelated tests?

**Comment by ADmad** [comment]
@hmic I restarted the failed build

File: tests/TestCase/Database/Type/DateTypeTest.php [hunk 2]
```diff
@@ -75,11 +75,11 @@ public function testToDatabase()
...
         $this->assertEquals($value, $result);
 
-        $date = new Time('2013-08-12');
+        $date = new Date('2013-08-12');
         $result = $this->type->toDatabase($date, $this->driver);
         $this->assertEquals('2013-08-12', $result);
 
-        $date = new Time('2013-08-12 15:16:18');
+        $date = new Date('2013-08-12 15:16:18');
         $result = $this->type->toDatabase($date, $this->driver);
```
Comment (markstory) [inline tests/TestCase/Database/Type/DateTypeTest.php]: We should probably keep a test usimg Time to make sure we don't make mistakes in the future.

**Comment by CauanCabral** [comment]
:+1:

**Comment by AD7six** [comment]
@hmic Travis regularly fails at random (or on time sensitive tests), we restart the builds when that happens unless it's a known failure.

As mark points out would be a good idea to keep the existing tests, and add new ones for using Date, otherwise :+1: .

**Comment by hmic** [comment]
Should i just duplicate (all) the tests with Time and Date?

**Comment by hmic** [comment]
I personally dont think there are any changes in the tests needed, as absolutely no functionality gets implemented here, just other default formatting options and the separation of datetime and date only formatting.

**Comment by AD7six** [comment]
> I personally dont think there are any changes in the tests needed

Well, there are now no checks that you get a `YYYY-MM-DD` back if you use the Time class. Any bug which changes that could go unnoticed.

**Comment by hmic** [comment]
These checks are (still) in the DateTimeTypeTest.php, not?
Maybe i should add more tests there then?

**Comment by markstory** [comment]
@hmic Default formatting is part of the API contract this class maintains, which should have a test. If I changed the formats by accident would you consider it a regression?

**Comment by hmic** [comment]
I kept all tests but changed the class to Date, still dont understand which test you want to see from the Time class in the tests for the DateType class.

**Comment by markstory** [comment]
I would want to see tests that show:

* Default string format for Date.
* timeAgoInWords() cuts off at the day boundaries.
* nice() format is what is expected.

I would put these tests in a new test file.

**Comment by hmic** [comment]
Still unsure which new tests could be needed.

- The code adds no new functionality
- It just inherits already tested code and changes a default format of it
- This format is already tested in the original tests

Which could only lead to
- alter the original tests to skip this cases, which would be wrong, as the DateTime can knowingly beeing formated as date-only
- duplicating the corresponding tests in an additional DateTest class which would only lead to duplication, but ensuring that the default formatting stays intact

This is absolutely not needed, as you will want to change this default to a localelized one anyway

The added/needed test to ensure the default stays the same would look like:

assertEquals('YYYY-mm-dd', ....boilsdownto... 'YYYY-mm-dd')


What else am I missing?

**Comment by markstory** [comment]
The tests would be to ensure the defaults don't change. Imagine a hypothetical scenario where all the defaults define in the new Date class were removed. Would you consider that a regression? If so then we should include tests to ensure that regression doesn't happen.

**Comment by lorenzo** [comment]
`resetToStringFormat()` needs to be overridden as well

**Comment by lorenzo** [comment]
I'll move this to 3.1 as I think it has a lot of value, but lacks some work

**Comment by hmic** [comment]
You should better move it to 4.0 then, because it will be a BC break for the current formatting of date fields, even though the current formatting is borked...

**Comment by lorenzo** [comment]
@hmic not really, Date can extend Time so the signatures will be the same. We can also offer a small snippet to make all Date columns use the old behavior.

**Comment by hmic** [comment]
If I could come up with an implementation and proper tests, when whould be the deadline for the 3.0 release?

**Comment by markstory** [comment]
Probably not, but there won't be a long time period until 3.1

**Pull request closed without merging**